File: Programmers/[42888] 오픈채팅방/JFe/42888.py
from collections import defaultdict
def solution(record):
    answer, log = [], []
    state = defaultdict(bool)   # 유저별 현재 채팅방에 있는지 여부
    nick = defaultdict(str)     # 유저별 닉네임 현황
    # record -> dictionay 처리
    for r in record:
        temp = r.split(" ")
        if temp[0] == "Enter":
            state[temp[1]] = True
            nick[temp[1]] = temp[2]
            log.append([temp[0], temp[1]])
        elif temp[0] == "Leave":
            state[temp[1]] = False
            log.append([temp[0], temp[1]])
        else:
            nick[temp[1]] = temp[2]
    # 최종 메세지 출력
    for command, id in log:
        if command == "Enter":
            answer.append(nick[id] + "님이 들어왔습니다.")
        else:
            answer.append(nick[id] + "님이 나갔습니다.")
    return answer

File: Programmers/[42888] 오픈채팅방/KingDonggyu/42888.js
function solution(record) {
  const userMap = new Map();
  const inoutRecord = [];

  record.forEach((str) => {
    const splited = str.split(' ');
    if (splited[0] === 'Change' || splited[0] === 'Enter') {
      userMap.set(splited[1], splited[2]);
    }

    if (splited[0] !== 'Change') {
      inoutRecord.push({
        id: splited[1],
        action: splited[0],
      });
    }
  });

  const answer = [];

  inoutRecord.forEach(({ id, action }) => {
    const nickName = userMap.get(id);
    if (action === 'Enter') {
      answer.push(`${nickName}님이 들어왔습니다.`);
      return;
    }
    answer.push(`${nickName}님이 나갔습니다.`);
  });

  return answer;
}
